DYN-6052 Add node info event and API under DynamoServices package for ProtoCore

QilongTang opened the pull request:
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

- [x] Add node info event and API under DynamoServices package for ProtoCore so libG could raise info message instead of warning message.
- [x] Add mechanism for graph run to consume runtime infos and connect each of them with node.

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB 

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section** 


### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

pinzart90 commented on src/Engine/ProtoCore/RuntimeStatus.cs:
```diff
@@ -52,13 +52,31 @@ public struct WarningEntry
+            public Guid GraphNodeGuid;
+            public int AstID;
+            public string Filename;
+        }
```
I wish we didn’t have to expose more APIs from the engine level

QilongTang replied:
I did look at it but unfortunately because the current `public interface ILiveRunner` setup, the function involved needs to be public and also the return object as well. VS does throw errors if trying to make this internal

mjkkirschner replied:
Is this really a good candidate for a struct?

mjkkirschner replied:
https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

QilongTang replied:
@mjkkirschner I am not sure but maybe we could break these changes in 3.0. Right now I am just following the previous design for all these engine related code

QilongTang commented on src/DynamoCore/Graph/Workspaces/HomeWorkspaceModel.cs:
```diff
@@ -765,20 +765,33 @@ private void OnUpdateGraphCompleted(AsyncTask task)
+                    continue;
+                using (node.PropertyChangeManager.SetPropsToSuppress(nameof(NodeModel.ToolTipText), nameof(NodeModel.Infos), nameof(NodeModel.State)))
+                {
+                    node.Info(string.Join(Environment.NewLine, info.Value.Select(w => w.Message)));
```
This is the place where we would display node info returned from libG

QilongTang commented on src/Engine/ProtoCore/RuntimeStatus.cs:
```diff
@@ -99,6 +139,7 @@ public void ClearWarningForExpression(int expressionID)
             using (rwl.CreateWriteLock())
             {
                 warnings.RemoveAll(w => w.ExpressionID == expressionID);
+                infos.RemoveAll(w => w.ExpressionID == expressionID);
```
@aparajit-pratap I am not 100% sure about the usage of these function or if infos need to be reset. But I guess we want to clear the runtime infos every time graph runs. Please correct me if I was wrong.

aparajit-pratap replied:
This could be tricky. In the case of geometry scale info messages, we would want it to be associated with node execution, while in other contexts, it may just be a static info message that has nothing to do with the node execution, in which case this may not be applicable.

QilongTang commented on src/NodeServices/MessageEvents.cs:
```diff
@@ -24,6 +37,18 @@ public static void OnLogWarningMessage(string message)
+    public static class LogInfoMessageEvents
+    {
+        public static event LogInfoMessageEventHandler LogInfoMessage;
+        public static void OnLogInfoMessage(string message)
```
This is the new API ProtoGeometry will call to raise info instead of warning

QilongTang commented on src/Engine/ProtoCore/RuntimeStatus.cs:
```diff
@@ -196,11 +240,81 @@ public void LogWarning(Runtime.WarningID ID, string message, string filename, in
+#if DEBUG
+            if (runtimeCore.Options.Verbose)
+            {
+                Console.WriteLine(infoMsg);
```
@aparajit-pratap @pinzart90 @sm6srw @mjkkirschner Do you have a suggestion on if writing info would be as valuable as writing runtime warnings?

QilongTang replied:
I end up removing the logging since we may not care about graph info as much as a warning. We can add it back if we care

aparajit-pratap commented on src/Engine/ProtoCore/RuntimeStatus.cs:
```diff
@@ -52,13 +52,31 @@ public struct WarningEntry
+        {
+            internal Runtime.InfoID ID;
+            internal string Message;
+            internal int Line;
+            internal int Column;
```
I don't think we need Line and Column. These were originally used for warnings on a per line basis for code block node scripts.

aparajit-pratap reviewed (commented):
@QilongTang I tried an alternative approach using VMDataBridge in the referenced PR, please check it out. It avoids having to make these many changes at the VM level.

aparajit-pratap reviewed (commented):
@QilongTang I'm okay with going ahead with this PR instead of mine as I wasn't able to figure out how to clear Info messages using the VM Databridge approach. If you could address review comments made so far, that would be great.

QilongTang commented:
Addressed the comments, the unit test with an example of a large number node triggering info cant be tested without the change to libG side. So I may merge this first and have a follow-up PR to add unit test.

aparajit-pratap commented on src/Engine/ProtoCore/BuildStatus.cs:
```diff
@@ -97,6 +100,16 @@ public OutputMessage(string message)
         }
 
+        // A constructor for info messages without line and colume.
+        public OutputMessage(MessageType type, string message, string filePath)
```
pls make this internal

QilongTang replied:
Made internal

aparajit-pratap commented on src/NodeServices/MessageEvents.cs:
```diff
@@ -12,6 +12,19 @@ public LogWarningMessageEventArgs(string msg)
+    /// <summary>
+    /// EventArgs for LogInfoMessageEventHandler
+    /// </summary>
+    public class LogInfoMessageEventArgs : EventArgs
+    {
+        public string message { get; internal set; }
+
+        public LogInfoMessageEventArgs(string msg)
+        {
+            message = msg;
+        }
+    }
```
We can simply use `LogWarningMessageEventArgs`. For all practical purposes, this is the same as that class so this seems redundant.

QilongTang replied:
Will do if you prefer, originally I was going for more separate API.

QilongTang replied:
Done

aparajit-pratap reviewed (commented):
Few more comments.

QilongTang commented:
Since we do not have 3.0 build working yet, I will have to merge to master -> cherry-pick to 2.19 branch and then use a 2.19 beta nuget to test. Merging.
